File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
       int row= matrix.size();
       int col= matrix[0].size();
       vector <int> ans;

       int l=0;
       int r= col-1;
       int t= 0;
       int b= row-1;

       while(l<=r && t<=b){
            for (int i=l; i<=r ; i++){
                ans.push_back(matrix[t][i]);
            }
            t++;

            for (int i=t; i<=b ; i++){
                ans.push_back(matrix[i][r]);
            }
            r--;


            if (t<=b){
                for (int i=r; i>=l ; i--){
                ans.push_back(matrix[b][i]);
            }
            b--;
            }
            

            if (l<=r){
                for (int i=b; i>=t ; i--){
                ans.push_back(matrix[i][l]);
            }
            l++;
            }
            
        } 

        return ans;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        m= len(matrix)
        n= len(matrix[0])
        ans=[]

        l= 0
        r= n-1
        t=0
        b= m-1

        while (l<=r and t<=b):
            for i in range(l, r+1):
                ans.append(matrix[t][i])

            for i in range(t+1, b+1):
                ans.append(matrix[i][r])

            if (t<b):

                for i in range (r-1, l-1, -1):
                    ans.append(matrix[b][i])

            if(l<r):
                for i in range( b-1, t, -1):
                    ans.append(matrix[i][l])

            l+=1
            r-=1
            t+=1
            b-=1

        return ans
        